File: lib/habitat/os/arch_linux.ex
defmodule Habitat.OS.ArchLinux do
  alias Habitat.Container
  alias Habitat.PackageManager.Pacman

  @chaotic_aur_key "3056513887B78AEB"
  @chaotic_aur_keyring "https://cdn-mirror.chaotic.cx/chaotic-aur/chaotic-keyring.pkg.tar.zst"
  @chaotic_aur_mirror_list "https://cdn-mirror.chaotic.cx/chaotic-aur/chaotic-mirrorlist.pkg.tar.zst"
  @key_server "hkp://keyserver.ubuntu.com:80"

  def image(), do: "archlinux"

  def install(container, packages) do
    Pacman.install(container, packages)
  end

  def uninstall(container, packages) do
    Pacman.uninstall(container, packages)
  end

  def post_create(container) do
    add_chaotic_aur(container)
  end

  defp add_chaotic_aur(container) do
    {_, 0} = Container.cmd(container, ["sudo", "pacman-key", "--init"])

    {_, 0} =
      Container.cmd(
        container,
        ["sudo", "pacman-key", "--recv-key", @chaotic_aur_key, "--keyserver", @key_server]
      )

    {_, 0} = Container.cmd(container, ["sudo", "pacman-key", "--lsign-key", @chaotic_aur_key])

    {_, 0} =
      Container.cmd(
        container,
        ["sudo", "pacman", "--upgrade", "--noconfirm", @chaotic_aur_keyring]
      )

    {_, 0} =
      Container.cmd(
        container,
        ["sudo", "pacman", "--upgrade", "--noconfirm", @chaotic_aur_mirror_list]
      )

    pacman_conf = File.read!("/etc/pacman.conf")

    unless pacman_conf =~ ~r/chaotic-aur/ do
      string =
        """
        #{pacman_conf}
        [chaotic-aur]
        Include = /etc/pacman.d/chaotic-mirrorlist
        """

      File.write!("/tmp/pacman.conf", string)

      {_, 0} = Container.cmd(container, ["sudo", "cp", "/tmp/pacman.conf", "/etc/pacman.conf"])
    end
  end
end

File: lib/habitat/os/archlinux.rb
module Habitat
  module OS
    class ArchLinux
      def initialize(shell)
        @shell = shell

        @links = Links.new(shell)
        @locales = Locales.new(shell)
        @packages = Pacman.new(shell)
        @sources = Sources.new(shell)
      end

      def sync(spec)
        @locales.sync(spec)
        @sources.sync(spec)
        @packages.sync(spec)
        @links.sync(spec)
      end

      class Links
        def initialize(shell)
          @shell = shell
        end

        def sync(spec)
          spec.links.each do |link|
            destination =
              if File.directory?(link.destination)
                File.join(link.destination, File.basename(link.source))
              else
                link.destination
              end

            FileUtils.mkdir_p File.dirname(destination)

            if !File.exist?(destination)
              FileUtils.ln_s(link.source, destination)
            elsif File.symlink?(destination)
              next
            else
              warn "Skipping #{destination}: file exists"
            end
          end
        end
      end

      class Sources
        def initialize(shell)
          @shell = shell
        end

        def sync(spec)
          spec.sources.each do |source|
            switches = [
              ["--recv-key", source.key],
              ["--keyserver", source.keyserver]
            ]

            @shell.run(["pacman-key", *switches.flatten], sudo: true)
            @shell.run(["pacman-key", "--lsign-key", source.key], sudo: true)

            switches = [
              "--needed",
              "--noconfirm",
              "--upgrade"
            ]

            @shell.run(["pacman", *switches.flatten, *source.packages], sudo: true)

            pacman_conf = @shell.read("/etc/pacman.conf")

            next unless pacman_conf !~ /\[#{source.name}\]/

            write(
              "/etc/pacman.conf",
              "#{pacman_conf}\n\n[#{source.name}]\nInclude = /etc/pacman.d/#{source.mirrorlist}\n",
              sudo: true
            )
          end
        end
      end

      class Locales
        def initialize(shell)
          @shell = shell
        end

        def sync(spec)
          @shell.write "/etc/locale.gen", locale_gen(spec.locales), sudo: true
          @shell.write "/etc/locale.conf", locale_conf(spec.locales), sudo: true

          @shell.run "locale-gen", sudo: true
        end

        private def locale_gen(locales)
          locales.map { |l| "#{l} UTF-8" }.join("\n") + "\n"
        end

        private def locale_conf(locales)
          "LANG=#{locales.first}\n"
        end
      end

      class Pacman
        def initialize(shell)
          @shell = shell
        end

        def sync(spec)
          pkgs = spec.packages.filter { |pkg| pkg.source == :pacman }

          @shell.run(
            [
              "pacman",
              "--sync",
              "--refresh",
              "--needed",
              "--noconfirm",
              *pkgs.map(&:to_s)
            ],
            sudo: true
          )
        end
      end
    end
  end
end
